### ISSUE TITLE
Not working with CH340G USB-serial adapter

### ISSUE THREAD

**Issue by spth** [spth/OpenRabbit#26]
I've tried with a CH340G USB-to-serial adapter and the serial programming cable.

However, this fails:

~~~~
Reset Rabbit.
loaded 831 bytes from ../coldboot/coldload.bin
Sending initial loader.
Warning: Processor verification sequence failed!
sending 276 initial loader triplets
Error: Status line should be high after sending initial loader.
~~~~

The error is the same no matter if the new tcdrain() or old usleep() approach is used in rabbit.c

**Comment by spth** [comment]
Actually, I sometimes get the same error with a FT232R (i.e. an original Rabbit USB programming cable) when it is connected to the host via a hub. Probably should recheck the CH340G with vs. without hub.

**Comment by spth** [comment]
Looks like hub vs. no Hub makes no difference.

**Comment by spth** [comment]
Looking at the line using a logic analyzer, my CH340G-based USB-serial-converter reliably transmits at 380 baud when set to 2400 baud. So the primary loader never gets received by the Rabbit.

**Comment by tomlogic** [comment]
🤦‍♂️ So if you just used a terminal emulator, you'd see the same thing?

Possibly a driver error, but it'd be difficult to track down driver updates on something that old.  I wonder if there are drivers by other manufacturers of CH340G-based hardware that you could use instead.

**Comment by spth** [comment]
Yes, I can reproduce the problem using gtkterm. The converter works at 4800 baud and 9600 baud. But when set to 2400 baud or 1200 baud, it transmits at 380 baud instead. I use the driver that is in Linux, i.e. drivers/usb/serial/ch341.c. I guess I should get another adapter or two, to check if this is a general problem with CH34x, or just one bad converter.

**Comment by tomlogic** [comment]
There could be a bug in the baud rate calculation code in that driver.  If you have access to a Windows machine (or a Mac?), you might want to do some testing there.

You could also copy `ch341_set_baudrate_lcr()` and `ch341_get_divisor()` out of [that driver](https://github.com/torvalds/linux/blob/master/drivers/usb/serial/ch341.c), and create a dummy program that generates a table of calculated values for various baud rates to check their validity.

I'm not familiar with Linux device drivers, but I'm also seeing some code related to "quirks".  Is it possible to configure the driver to use `CH341_QUIRK_LIMITED_PRESCALER` (or something else like it) for your particular adapter?

```C
	/* Some devices require a lower base clock if ps < 3. */
	if (ps < 3 && (priv->quirks & CH341_QUIRK_LIMITED_PRESCALER))
		force_fact0 = true;
```

Just throwing out some ideas if you're interested in going down a non-OpenRabbit rabbit hole. 🐇

**Comment by spth** [comment]
I just tried with a Raspi 4 running FreeBSD 14.2, and see the same problem there. However, the serial adapter works at 2400 baud on a Mac Mini G4 running OpenBSD 7.6.

FreeBSD bug report: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=287509
Linux bug report: https://bugzilla.kernel.org/show_bug.cgi?id=220225

**Comment by spth** [comment]
I see the same problem using a second such adaper.
